docs/source/notes/gradcheck.rst

```diff
@@ -0,0 +1,223 @@
+"""""""""""""""""""""""""""""""""""""""
+
+We first consider how to compute :math:`s` with a numerical method. To do so, keeping in mind that we're considering :math:`g: \mathcal{C}^N \to \mathcal{R}^M, z \to y` with :math:`z = a + i b`, and that :math:`CW = \frac{1}{2} * (\frac{\partial y}{\partial a} + i \frac{\partial y}{\partial b})`,  we rewrite it as follows:
+
+.. math::
+    \begin{aligned}
+        s &= 2 * v^T (real(CW) ur + i * imag(CW) ui) \\
+          &= 2 * v^T (\frac{1}{2} * \frac{\partial y}{\partial a} ur + i * \frac{1}{2} * \frac{\partial y}{\partial b} ui) \\
+          &= v^T (\frac{\partial y}{\partial a} ur + i * \frac{\partial y}{\partial b} ui) \\
+          &= v^T ((\frac{\partial y}{\partial a} ur) + i * (\frac{\partial y}{\partial b} ui))
+    \end{aligned}
+
+In this formula, we can see that :math:`\frac{\partial y}{\partial a} ur` and :math:`\frac{\partial y}{\partial b} ui` can be evaluated the same way as the fast version for the real-to-real case.
+Once these real-valued quantities have been computed, we can reconstruct the complex vector on the right side and do a dot product with the real-valued :math:`v` vector.
+
+Fast complex input analytical evaluation
+""""""""""""""""""""""""""""""""""""""""
+
+For the analytical case, things are simpler and we rewrite the formula as:
+
+.. math::
+    \begin{aligned}
+        s &= 2 * v^T (real(CW) ur + i * imag(CW) ui) \\
+          &= v^T real(2 * CW) ur + i * v^T imag(2 * CW) ui) \\
+          &= real(v^T (2 * CW)) ur + i * imag(v^T (2 * CW)) ui
+    \end{aligned}
+
+We can thus use the fact that the backward mode AD provides us with an efficient way to compute :math:`v^T (2 * CW)` and then perform a dot product of the real part with :math:`ur` and the imaginary part with :math:`ui` before reconstructing the final complex scalar :math:`s`.
+
+Why not use a complex :math:`u`
+"""""""""""""""""""""""""""""""
+
+At this point, you might be wondering why we did not select a complex :math:`u` and just performed the reduction :math:`2 * v^T CW u'`.
+The problem is that when doing the numerical evaluation, considering :math:`u' = ur' + i ui'`, we would need to compute:
+
+.. math::
+    \begin{aligned}
+        2*CW u' &= (dy/da + i dy/db)(ur' + i ui') \\
+                &= dy/da ur' + i dy/da ui' + i dy/db ur' - dy/db ui'
+    \end{aligned}
+
+Which would require four evaluations of real-to-real finite difference (twice as much compared to the approached proposed above).
+Since this approach does not have more degrees of freedom (same number of real valued variables) and we try to get the fastest possible evaluation here, we use the other formulation above.
+
+
+Fast gradcheck for functions with outputs in :math:`\mathcal{C}^M`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Just like in the slow case, we consider two real-valued function and use the appropriate rule from above for each function.
+
+Gradgrad check implementation
```

Comment:
gradgrad check -> gradgradcheck?

Reply:
I see you also have gradgrad check below, not sure if you meant to do so

Reply:
Yes moving to one word to be consistent with how we use "gradcheck"